backend/peephole/peephole_rules.ml

```diff
@@ -0,0 +1,182 @@
+(* CR-someday: see whether the `-4` can be dropped. *)
+[@@@ocaml.warning "+a-29-40-41-42-4"]
+
+open! Peephole_utils
+
+(** Logical condition for simplifying the following case:
+    {|
+    mov x, y
+    mov y, x
+    |}
+
+    In this case, the second instruction should be removed *)
+
+let remove_useless_mov (cell : Cfg.basic Cfg.instruction DLL.cell) =
+  match get_cells cell 2 with
+  | [fst; snd] -> (
+    let fst_val = DLL.value fst in
+    let snd_val = DLL.value snd in
+    match fst_val.desc with
+    | Op (Move | Spill | Reload) -> (
+      let fst_src, fst_dst = fst_val.arg.(0), fst_val.res.(0) in
+      match snd_val.desc with
+      | Op (Move | Spill | Reload) ->
+        let snd_src, snd_dst = snd_val.arg.(0), snd_val.res.(0) in
+        if are_equal_regs fst_src snd_dst && are_equal_regs fst_dst snd_src
+        then (
+          DLL.delete_curr snd;
+          Some (prev_at_most go_back_const fst))
+        else None
+      | _ -> None)
+    | _ -> None)
+  | _ -> None
+
+(** Logical condition for simplifying the following case:
+  {|
+    <op 1> const1, r
+    <op 2> const2, r
+  |}
+
+  to:
+  {|
+    <op 1> (const1 <op 2> const2), r
+  |}
+
+   Where <op 1> and <op 2> can be any two binary operators that are associative
+   and const1 and const2 are immediate values. *)
+
+let are_compatible op1 op2 imm1 imm2 =
+  match (op1 : Mach.integer_operation), (op2 : Mach.integer_operation) with
+  (* Folding two bitwise operations such as (AND, OR, XOR) should never produce
+     an overflow. *)
+  | Mach.Iand, Mach.Iand ->
+    bitwise_overflow_assert imm1 imm2 ( land );
+    Some (Mach.Iand, imm1 land imm2)
+  | Ior, Ior ->
+    bitwise_overflow_assert imm1 imm2 ( lor );
+    Some (Mach.Ior, imm1 lor imm2)
+  | Ixor, Ixor ->
+    bitwise_overflow_assert imm1 imm2 ( lxor );
+    Some (Mach.Ixor, imm1 lxor imm2)
+  (* For the following three cases we have the issue that in some situations,
+     one or both immediate values could be out of bounds, but the result might
+     be within bounds (e.g. imm1 = -4 and imm2 = 65, their sum being 61), in
+     that case we still do the folding. This should not happen at all since the
+     immediate values should always be within the bounds [0, Sys.int_size]. *)
```

Comment:
The code and the comment do not
look in sync. Indeed, I think we
should (conservatively) not simplify
in the `-4`, `65` case but we do
below.